### PR TITLE
Provide an API to supersede System.getenv

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

JVM stores all environment variables in an internal `UnmodifiableMap` and doesn't allow us to modify it. Previously we did a hackery for Gradle daemon - reflectively modify these environment variables, which is more and more difficult after Java 9. Aiming at superseding `System.getenv`, this PR provides a new API to get environment variables on daemon side.

Basically, what this PR does is:

- Provides a new `GradleSystem` to fetch current "Gradle-managed" environment variables.
- When daemon resets environment variable, it also reset these values in `GradleSystem`.
- When a new process is forked by daemon, `GradleSystem`'s environment variables are also passed to the forked process.

### CODE REVIEW

File: subprojects/base-services/src/main/java/org/gradle/api/GradleSystem.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import javax.annotation.Nullable;
+import java.util.Map;
+
+/**
+ * Gradle environmental variable collection.
+ *
+ * @since 4.10
+ */
+@Incubating
+public class GradleSystem {
```
Comment (oehme) [inline subprojects/base-services/src/main/java/org/gradle/api/GradleSystem.java]: Did you consider an instance method like `StartParameter#getEnv`? What made you choose a static method instead?

Comment (blindpirate) [reply subprojects/base-services/src/main/java/org/gradle/api/GradleSystem.java]: I think the transition `System.getenv` -> `GradleSystem.getenv` would make it better for users to understand and use.

**Comment by oehme** [comment]
The TBD items above don't have to be in this pull request. Doing automatic code rewriting/warning is not critical, though a big bonus. If we do that, let's do it separately so we can get this one merged soon.

We just need some documentation, probably in the "build environment" chapter. Also we need to make sure that setting Gradle properties through `ORG_GRADLE_PROJECT_propertyName`-style variables now works again on Java 9+.

File: subprojects/base-services/src/main/java/org/gradle/api/GradleSystem.java (hunk 1 above)
Comment (oehme) [reply subprojects/base-services/src/main/java/org/gradle/api/GradleSystem.java]: 👍

Comment (melix) [reply subprojects/base-services/src/main/java/org/gradle/api/GradleSystem.java]: On the other hand it makes the API rather unsafe, and prevents us from doing smarter things in the future like providing different environments to workers. I'd make it an instance API.

Comment (oehme) [reply subprojects/base-services/src/main/java/org/gradle/api/GradleSystem.java]: It doesn't prevent us from providing different environments to workers. The worker API already has a way to mutate the environment on a per-worker basis.

**Comment by melix** [comment]
What I mean is that a static variable is shared by the class loader. This
inherently removes a lot of flexibility. I don't see any advantage in
restricting us like this.

Le mer. 18 juil. 2018 à 14:32, Stefan Oehme <notifications@github.com> a
écrit :

**Comment by blindpirate** [comment]
@melix Which kind of API form do you expect?

**Comment by oehme** [comment]
@melix I don't know what kind of use case you have in mind, can you please elaborate?

**Comment by oehme** [comment]
@blindpirate I had a chat with Adam about this and we agreed to take a stab at a public `BuildOption` API which allows users to map environment variables to build options instead of accessing a static variable. Assinging this to me.

**Comment by blindpirate** [comment]
Okay, thanks!

**Comment by wolfs** [comment]
@oehme Are we still going to provide an API for environment variables from the Gradle side? Is there an issue for doing so?

**Comment by oehme** [comment]
@wolfs That's covered by https://github.com/gradle/gradle/issues/2885

**Pull request closed without merging**